opm/simulators/flow/BlackoilModelEbos.cpp

```diff
@@ -0,0 +1,130 @@
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/simulators/flow/BlackoilModelEbos.hpp>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <cmath>
+#include <iomanip>
+
+namespace Opm {
+namespace detail {
+
+template<class Scalar>
+ConvergenceReport createConvergenceReport(const int iteration,
+                                          const double reportTime,
+                                          const double tol_mb,
+                                          const double tol_cnv,
+                                          const double maxResidualAllowed,
+                                          const bool terminal_output,
+                                          const std::vector<std::string>& names,
+                                          const std::vector<Scalar>& CNV,
+                                          const std::vector<Scalar>& mass_balance_residual,
+                                          const std::string& terminal_header)
+{
+    // Create convergence report.
+    ConvergenceReport report{reportTime};
+    using CR = ConvergenceReport;
+    const int numComp = CNV.size();
+    for (int compIdx = 0; compIdx < numComp; ++compIdx) {
+        double res[2] = { mass_balance_residual[compIdx], CNV[compIdx] };
+        CR::ReservoirFailure::Type types[2] = { CR::ReservoirFailure::Type::MassBalance,
+                                                CR::ReservoirFailure::Type::Cnv };
+        double tol[2] = { tol_mb, tol_cnv };
+        for (int ii : {0, 1}) {
+            if (std::isnan(res[ii])) {
+                report.setReservoirFailed({types[ii], CR::Severity::NotANumber, compIdx});
+                if (terminal_output) {
+                    OpmLog::debug("NaN residual for " + names[compIdx] + " equation.");
+                }
+            } else if (res[ii] > maxResidualAllowed) {
+                report.setReservoirFailed({types[ii], CR::Severity::TooLarge, compIdx});
+                if (terminal_output) {
+                    OpmLog::debug("Too large residual for " + names[compIdx] + " equation.");
+                }
+            } else if (res[ii] < 0.0) {
+                report.setReservoirFailed({types[ii], CR::Severity::Normal, compIdx});
+                if (terminal_output) {
+                    OpmLog::debug("Negative residual for " + names[compIdx] + " equation.");
+                }
+            } else if (res[ii] > tol[ii]) {
+                report.setReservoirFailed({types[ii], CR::Severity::Normal, compIdx});
+            }
+            report.setReservoirConvergenceMetric(types[ii], compIdx, res[ii]);
```

Comment:
this was originally not done for subdomains but i think it was just an oversight.